src/Exports/DonorsExport.php

```diff
@@ -0,0 +1,212 @@
+
+        if($this->searchBy === 'donor') {
+            if( $this->startDate && $this->endDate ) {
+                $donorQuery->whereBetween('donors.date_created', $this->startDate, $this->endDate );
+            } elseif( $this->startDate ) {
+                $donorQuery->where('donors.date_created', $this->startDate, '>=');
+            } elseif( $this->endDate ) {
+                $donorQuery->where('donors.date_created', $this->endDate, '<');
+            }
+        }
+        else {
+            if( $this->startDate && $this->endDate ) {
+                $donationQuery->whereBetween('donations.post_date', $this->startDate, $this->endDate );
+            } elseif( $this->startDate ) {
+                $donationQuery->where('donations.post_date', $this->startDate, '>=');
+            } elseif( $this->endDate ) {
+                $donationQuery->where('donations.post_date', $this->endDate, '<');
+            }
+        }
+
+        $donorQuery->joinRaw( "JOIN ({$donationQuery->getSQL()}) AS sub ON donors.id = sub.donorId" );
+
+        $results = $donorQuery->getAll(ARRAY_A);
+
+        $exportData = array_map([$this, 'mapDonorColumnNames'], $results);
+
+        if( $this->shouldIncludeAddress() ) {
+            $exportData = array_map([$this, 'mapDonorAddress'], $exportData);
+        }
+
+        $exportData = array_map([$this, 'filterIncludedColumns'], $exportData );
+        $exportData = array_map([$this, 'flattenAddressColumn'], $exportData );
+
+        return $this->filterExportData( $exportData );
+    }
+
+    /**
+     * @param array $donorData
+     * @return array
+     */
+    protected function mapDonorColumnNames( $donorData )
+    {
+        return [
+            'full_name'          => $donorData[ 'name' ],
+            'email'              => $donorData[ 'email' ],
+            'userid'             => $donorData[ 'user_id' ],
+            'donations'          => $donorData[ 'purchase_count' ],
+            'donation_sum'       => $donorData[ 'purchase_value' ],
+        ];
+    }
+
+    /**
+     * @return bool
+     */
+    protected function shouldIncludeAddress()
+    {
+        return isset( $this->posted_data[ 'give_export_columns' ][ 'address' ] );
+    }
+
+    /**
+     * @param array $columnarData
+     * @return array
+     */
+    protected function flattenAddressColumn( $columnarData )
+    {
+        return $this->flattenColumn( $columnarData, 'address' );
+    }
+
+    /**
+     * @param array $columnarData
+     * @param string $columnName
+     * @return array
+     */
+    protected function flattenColumn( $columnarData, $columnName )
+    {
+        if( isset( $columnarData[ $columnName ])) {
+            $columnarData = array_merge( $columnarData, $columnarData[ $columnName ] );
+            unset( $columnarData[ $columnName ] );
+        }
+        return $columnarData;
+    }
+
+    /**
+     * @TODO N+1 when querying for each donor's address.
```

Comment:
@jonwaldstein @alaca is the attach meta stuff in the query builder ready to go?

Reply:
@jonwaldstein @kjohnson I don't think `attachMeta` will work for the address fields, because `attachMeta` requires you to specify which meta keys to include. Here's a slice from `give_donormeta`:
![image](https://user-images.githubusercontent.com/7819986/157312919-4e45eae5-4afe-4b5e-90ef-0b3ebf7a4007.png)

Since we don't know how many different addresses there are, we'd probably just want a separate query, with a clause like `whereLike('meta_key', '_give_donor_address_billing')`.

Reply:
oof yeah I just realized donors can have more than 1 address.  still have to add that to the models..

Let's see what the wizard @alaca thinks.  It would be really nice if we could do something like 
```
->attachmeta('give_donormeta',
  'ID',
  'donor_id',
  ['_give_donor_address_*]
)
 ```
 
Either way, I think we'll need some kind of a join from the main query to prevent the n+1 issue

Reply:
Looking at the implementation of `give_get_donor_address()`, it looks like we are assuming the first (key of `0`) address. That should be safe to query against and would not be a breaking change.

Reply:
While Give supports multiple donor billing addresses, there is only one *main* billing address per donor.

Reply:
Trying `attachMeta` locally, and it looks like it will work. Just unclear on how we should be referencing the foreign key, which is ambiguous if you don't specify the corresponding table.

Reply:
So far, it looks like using the known alias for the foreign key works.

```php
    $donorQuery->attachMeta('give_donormeta',
        'donors.ID',
        'donor_id',
        '_give_donor_address_billing_line1_0',
        '_give_donor_address_billing_line2_0',
        '_give_donor_address_billing_city_0',
        '_give_donor_address_billing_state_0',
        '_give_donor_address_billing_country_0',
        '_give_donor_address_billing_zip_0',
    );
```

Reply:
@jonwaldstein @LMNTL see https://github.com/impress-org/givewp/pull/6236/commits/9fa214001f716aba72f887a5159b1003673c516c

Reply:
After giving it some thought, a column alias is required by `attachMeta`, even if we don't explicitly specify it, `attachMeta` will add it anyway to reference the `meta_value` column.

We could probably support `['give_donor_address*']` by using `UNION`, but then we still have a problem with aliases because there is no way to add them dynamically from result values - which makes sense. 

I'd like to explore this a bit more, maybe I can think of something. I mean, having something like `['give_donor_address*']` would be awesome.

Reply:
The only way how we could pull this off is by creating a temporary table that contains all the meta keys, and then adding those keys to the `attachMeta`. This could work 😃